### PR TITLE
Dragging a node under the last item of the tree raises it a level

### LINKED ISSUE #34033: Problem placing node in tree with drag and drop
**Godot version:**
3.2 beta 1 mono

**OS/device including version:**
win10 64 home

**Issue description:**
The problem is that a node that is the son of another node cannot be located off of it.
To place it below you have to take the node out of the father.

In this GIF you can see that to locate the node "key_door2" under (not inside) of "key_door" I have to:
1st place "key_door2" above "key_door"
2nd now you can put "key_door2" under "key_door"

In conclusion, a child node cannot relocate to a lower position (not inside it) than the parent node.

GIF

**Steps to reproduce:**
Make a node a relative of another and then try to locate the innermost node just below its parent node, for this you will have to break the parent / child relationship.

**Minimal reproduction project:**
I do not think it's necessary.

### PR DESCRIPTION
**Pull request by giarve**
A children node couldn't be taken out of the father without scrolling back to the parent node icon
and thus making very troublesome to rearrange a children to be a sibling of its parent if the list of the nodes was long.

By adding an extra dragging line at the end of any children tree we can ensure that the user can move without effort a children up to parent's level even if the parent is many nodes away.

dragging

If needed, spacing or thickness for the new line can be increased.

EDIT: As requested, thickness has been increased from 1 to 2.

Fixes #34033.

### CODE REVIEW

**Comment by marcospb19** [comment]
It is possible to make the lines at the bottom thicker, like 2 or 3 pixels? It looks very hard to aim.

And for this to be merged I also think that aiming below everything should bring the node to the top level (below the root). This PR currently does not implement it.

**Comment by Zireael07** [comment]
+1 to making lines thicker.

File: scene/gui/tree.cpp [hunk 1]
```diff
@@ -3512,7 +3517,33 @@ TreeItem *Tree::_find_item_at_pos(TreeItem *p_item, const Point2 &p_pos, int &r_
 
 			return nullptr;
 		} else {
-			pos.y -= h;
+			if (p_item != root && !p_item->is_collapsed() && !p_item->get_children() && !p_item->get_next() && p_item->get_parent() != root && pos.y < h + cache.vseparation // pos.y < compute_item_height() + 2*vseparation
```
Comment (KoBeWi) [inline scene/gui/tree.cpp:3520]:
```suggestion
			if (p_item != root && !p_item->is_collapsed() && !p_item->get_first_child() && !p_item->get_next() && p_item->get_parent() != root && pos.y < h + cache.vseparation // pos.y < compute_item_height() + 2*vseparation
```
This was renamed.

Also you should remove the commented-out code.

**Comment by akien-mga** [comment]
Is this still relevant? The issue was closed as fixed in 3.5.

**Comment by KoBeWi** [comment]
The PR seems to do more than fixing the issue. Needs rebase though.

**Comment by pafuent** [comment]
@giarve, @akien-mga and @KoBeWi 
This could be closed, it was implemented in another PR
NodeDrag

**Comment by giarve** [comment]
> @giarve, @akien-mga and @KoBeWi This could be closed, it was implemented in another PR NodeDrag [ ](https://private-user-images.githubusercontent.com/6979945/377436598-3b7026cd-ecdc-438f-bab1-e97fa4857f55.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxNjQxNjgsIm5iZiI6MTcyOTE2Mzg2OCwicGF0aCI6Ii82OTc5OTQ1LzM3NzQzNjU5OC0zYjcwMjZjZC1lY2RjLTQzOGYtYmFiMS1lOTdmYTQ4NTdmNTUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTdUMTExNzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGQxYjU4N2FlNzk1YmExMDRiNjg4MTZiMDI3NDdmYTRhNzg1NDYwYmU4M2ZhNDZjN2U2ZmI4ZDAxOGM5NWZlNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.E7uBHpFYgSIozHPYVHV4e5Ve-gAJa4-HIZmOCQSgmhY) NodeDrag [ ](https://private-user-images.githubusercontent.com/6979945/377436598-3b7026cd-ecdc-438f-bab1-e97fa4857f55.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjkxNjQxNjgsIm5iZiI6MTcyOTE2Mzg2OCwicGF0aCI6Ii82OTc5OTQ1LzM3NzQzNjU5OC0zYjcwMjZjZC1lY2RjLTQzOGYtYmFiMS1lOTdmYTQ4NTdmNTUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTdUMTExNzQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGQxYjU4N2FlNzk1YmExMDRiNjg4MTZiMDI3NDdmYTRhNzg1NDYwYmU4M2ZhNDZjN2U2ZmI4ZDAxOGM5NWZlNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.E7uBHpFYgSIozHPYVHV4e5Ve-gAJa4-HIZmOCQSgmhY)

Can you check which PR is it? I cannot believe there are two PRs doing the same and they are not referenced at all...

**Comment by pafuent** [comment]
@giarve before posting my comment I did a quick search of that PR but I had no luck. But as you can see on the GIF that I attached, the behavior is there (unless there is something that I didn't get it properly)
I did the test on the latest master and I just also tested on 4.3 stable and it's also implemented there.

**Comment by giarve** [comment]
Well... maybe we will never find out 😂 . I used copilot to search through the PRs and nothing pops. It is weird because the behavior is exactly the same.

However, it's been 4 years since this PR, and it is solved somehow. 

Closing.

**Pull request closed without merging**